### ISSUE TITLE
vui_paremeters parsing in an AVC file

### ISSUE THREAD

**Issue by VincentEmond** [MediaArea/MediaInfoLib#1409]
Hi,

I am using v21.03 of mediainfolib and mediainfo v21.03 on ubuntu 20.04. I am building my own mp4/avc parser and have been using the mediainfo CLI to compare results. Now everything matched correctly up until, the parsing of the seq_parameter_set NAL unit, more specifically the timing info section of vui_parameters.

This is the avcC box I am parsing:

> 00000217: 0000 0032 6176 6343 0164 001e ffe1 001a  ...2avcC.d......
> 00000227: 6764 001e acd9 40a0 2ff9 7011 0000 0300  gd....@./.p.....
> 00000237: 0100 0003 003c 0f16 2d96 0100 0568 ebec  .....<..-....h..
> 00000247: b22c  

This is the result I get from media info:

> 232            timing_info_present_flag (9 bytes)
> 232             timing_info_present_flag:   Yes
> 233             num_units_in_tick:          1 (0x00000001) - (32 bits)
> 237             time_scale:                 60 (0x0000003C) - (32 bits)
> 23B             fixed_frame_rate_flag:      No

This is the result I get from my script:

> Take 32 bits at offset 0x233 = num_units_in_tick: 768 (0x00000300)
> Take 32 bits at offset 0x237 = time_scale: 16 777 219 (0x01000003)

I am using **_ISO/IEC 14496-10:2020 (9th edition) Annex E - E.1.1 p.446_**. Is this a bug or is there something I misread from the standard? I don't understand how you extract 1 from the bitstream at offset 0x233 and 0x3C at offset 0x237

I have attached the file I am parsing. The command line I am using is `mediainfo --details=1 [file]`.

https://user-images.githubusercontent.com/9730015/122077242-919c6b00-cdc9-11eb-9e5b-8efa4fc57a2c.mp4

**Comment by VincentEmond** [comment]
Not sure if attaching the file worked correctly, here it is in base64:

> AAAAGGZ0eXBpc281AAACAGlzbzZtcDQxAAADD21vb3YAAABsbXZoZAAAAAAAAAAAAAAAAAAAA+gAAAAAAAEAAAEAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAIRdHJhawAAAFx0a2hkAAAAAwAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAQAAAAAKAAAABaAAAAAAAJGVkdHMAAAAcZWxzdAAAAAAAAAABAAAAAAAABAAAAQAAAAABiW1kaWEAAAAgbWRoZAAAAAAAAAAAAAAAAAAAPAAAAAAAVcQAAAAAAC1oZGxyAAAAAAAAAAB2aWRlAAAAAAAAAAAAAAAAVmlkZW9IYW5kbGVyAAAAATRtaW5mAAAAFHZtaGQAAAABAAAAAAAAAAAAAAAkZGluZgAAABxkcmVmAAAAAAAAAAEAAAAMdXJsIAAAAAEAAAD0c3RibAAAAKhzdHNkAAAAAAAAAAEAAACYYXZjMQAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAKAAWgASAAAAEgAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABj//wAAADJhdmNDAWQAHv/hABpnZAAerNlAoC/5cBEAAAMAAQAAAwA8DxYtlgEABWjr7LIsAAAAEHBhc3AAAAABAAAAAQAAABBzdHRzAAAAAAAAAAAAAAAQc3RzYwAAAAAAAAAAAAAAFHN0c3oAAAAAAAAAAAAAAAAAAAAQc3RjbwAAAAAAAAAAAAAAKG12ZXgAAAAgdHJleAAAAAAAAAABAAAAAQAAAAAAAAAAAAAAAAAAAGJ1ZHRhAAAAWm1ldGEAAAAAAAAAIWhkbHIAAAAAAAAAAG1kaXJhcHBsAAAAAAAAAAAAAAAALWlsc3QAAAAlqXRvbwAAAB1kYXRhAAAAAQAAAABMYXZmNTguMjkuMTAw

**Comment by JeromeMartinez** [comment]
As a quick check it seems that it is in the middle of a emulation_prevention_three_byte (0x000003), did you removed the emulation_prevention_three_byte stuff (7.3.1 in spec) during your parsing?

**Comment by VincentEmond** [comment]
I did not, I believe that was what I was missing thanks for the quick response!

**Comment by VincentEmond** [comment]
Actually I checked again and I DID implement that part too. However I found a bug in my code. When implementing the `next_bits()` function I used the `io.bufferedreader.peek(size)` function in python. Even though I called it with `peek(3)` to check for the `emulation_prevention_three_byte` the function returned the whole buffer, not three bytes. I implemented my own peek function and it's working fine now. I get the same result as mediainfo. If anyone is interested: [https://stackoverflow.com/questions/24474687/io-bufferedreader-peek-function-returning-all-the-text-in-the-buffer](https://stackoverflow.com/questions/24474687/io-bufferedreader-peek-function-returning-all-the-text-in-the-buffer) don't fall into the same trap as I did! Hours of debugging and manual parsing (good old pen and paper) for a stupid little detail. Thanks again for pointing me in the right direction, your comment was spot on!
